spec/unit/rom/relation/lazy/eager_load_spec.rb

```diff
@@ -0,0 +1,128 @@
+require 'spec_helper'
+
+describe ROM::Relation::Lazy, '#eager_load' do
+  include_context 'users and tasks'
+
+  before do
+    setup.relation(:users) do
+      def by_name(name)
+        restrict(name: name)
+      end
+    end
+
+    setup.relation(:tasks) do
+      def for_users(users)
+        names = users.map { |user| user[:name] }
+        restrict { |task| names.include?(task[:name]) }
+      end
+    end
+
+    setup.relation(:tags) do
+      forward :map
+
+      def for_tasks(tasks)
+        titles = tasks.map { |task| task[:title] }
+        restrict { |tag| titles.include?(tag[:task]) }
+      end
+
+      def for_users(users)
+        user_tasks = tasks.for_users(users)
+
+        for_tasks(user_tasks).map { |tag|
+          tag.merge(user: user_tasks.detect { |task| task[:title] == tag[:task] }[:name])
+        }
+      end
+    end
+
+    setup.repositories[:default].dataset(:tags).insert(task: 'be cool', name: 'red')
+    setup.repositories[:default].dataset(:tags).insert(task: 'be cool', name: 'green')
+  end
+
+  let(:users) { rom.relation(:users) }
+  let(:tasks) { rom.relation(:tasks) }
+  let(:tags) { rom.relation(:tags) }
+
+  let(:map_users) {
+    proc { |users, tasks|
+      users.map { |user|
+        user.merge(tasks: tasks.select { |task| task[:name] == user[:name] })
+      }
+    }
+  }
+
+  let(:map_tasks) {
+    proc { |(tasks, children)|
+      tags = children.first
+
+      tasks.map { |task|
+        task.merge(tags: tags.select { |tag| tag[:task] == task[:title] })
+      }
+    }
+  }
+
+  let(:map_user_with_tasks_and_tags) {
+    proc { |users, (tasks, tags)|
+      users.map { |user|
+      user_tasks = tasks.select { |task| task[:name] == user[:name] }
+
+      user_tags = tasks.flat_map { |task|
+        tags.select { |tag| tag[:task] == task[:title] }
+      }
+
+      user.merge(
+        tasks: user_tasks,
+        tags: user_tags
+      )
+    }
+    }
+  }
+
+  let(:map_user_with_tasks) {
+    proc { |users, children|
+      map_users[users, map_tasks[children.first]]
+    }
+  }
+
```

Comment:
Just noting here that there’s quite a lot of setup logic, before we reach the first test expectation, which makes me wonder if the test is testing these `:let` blocks as much as the actual eager load API.